src/IssueViz/IssueVisualizationControl/ViewModels/IssueVisualizationViewModel.cs

```diff
@@ -0,0 +1,207 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Shell.Interop;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Analysis;
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+using SonarLint.VisualStudio.IssueVisualization.Selection;
+
+namespace SonarLint.VisualStudio.IssueVisualization.IssueVisualizationControl.ViewModels
+{
+    internal sealed class IssueVisualizationViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly IAnalysisIssueSelectionService selectionEvents;
+        private readonly IVsImageService2 vsImageService;
+        private readonly IRuleHelpLinkProvider ruleHelpLinkProvider;
+        private readonly ILogger logger;
+
+        private IAnalysisIssueVisualization currentIssue;
+        private IAnalysisIssueFlowVisualization currentFlow;
+        private LocationListItem currentLocation;
+
+        private bool isBindingUpdatedOutsideOfControl;
+
+        public IssueVisualizationViewModel(IAnalysisIssueSelectionService selectionEvents, IVsImageService2 vsImageService, IRuleHelpLinkProvider ruleHelpLinkProvider, ILogger logger)
+        {
+            this.selectionEvents = selectionEvents;
+            this.vsImageService = vsImageService;
+            this.ruleHelpLinkProvider = ruleHelpLinkProvider;
+            this.logger = logger;
+
+            selectionEvents.SelectedIssueChanged += SelectionEvents_SelectedIssueChanged;
+            selectionEvents.SelectedFlowChanged += SelectionEventsOnSelectedFlowChanged;
+            selectionEvents.SelectedLocationChanged += SelectionEvents_SelectedLocationChanged;
+        }
+
+        public string Description => currentIssue?.Issue?.Message;
+
+        public string RuleKey => currentIssue?.Issue?.RuleKey;
+
+        public string RuleHelpLink => string.IsNullOrEmpty(RuleKey) ? null : ruleHelpLinkProvider.GetHelpLink(RuleKey);
+
+        public AnalysisIssueSeverity Severity => currentIssue?.Issue?.Severity ?? AnalysisIssueSeverity.Info;
+
+        public IAnalysisIssueVisualization CurrentIssue
+        {
+            get => currentIssue;
+            private set
+            {
+                currentIssue = value;
+                OnPropertyChanged(null);
```

Comment:
This raises a notification that all properties in the view model have changed. Have the current flow and current locations been updated at this point? It doesn't look like they have.
It looks as if changing the selected issue will cause the service to raise multiple events, which in turn will trigger multiple unnecessary refresh passes of the UI.

I think the root of the problem is with the service: it starts raising events before the model is stable.

Reply:
Hm, perhaps we can change the service to raise each time a single event with all the changed data, so for example:
1. When issue changes -> the service will raise IssueChanged event that will contain Issue, Flow and Location (instead of raising IssueChanged, FlowChanged and LocationChanged).
2. When flow changes -> the service will raise FlowChanged event that will contain Flow and Location (instead of raising FlowChanged and LocationChanged).

WDYT?

Reply:
I think it would be better if the service didn't start raising events until it had finished making state changes. That's confusing to work with - if you handle the `IssueChanged` event, you shouldn't ask the service what the currently selected flow is, since it will belong to a different issue.

The data is hierarchical, so it isn't possible to change the flow without changing the location, or the issue without changing the flow. One option would be to make all of the state changes, then raise a single event that specifies which level of the hierarchy changed - All, FlowAndLocation, LocationOnly.

Reply:
Good point about the flow/location being wrong when listening to IssueChanged event.
I'll make the changes in a separate PR.